File: app/src/main/java/com/jjf/template/ui/UserViewModel.java
package com.jjf.template.ui;

import com.jjf.template.bean.Resource;
import com.jjf.template.bean.User;
import com.jjf.template.repository.UserRepository;
import com.jjf.template.util.lifecycle.AbsentLiveData;

import javax.inject.Inject;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

/**
 * @author xj
 * date: 18-11-8
 * description :
 */

public class UserViewModel extends ViewModel {

    private  UserRepository mUserRepository;
    private MutableLiveData<String> loginData = new MutableLiveData<>();

    @Inject
    public UserViewModel(UserRepository userRepository) {
        this.mUserRepository = userRepository;
    }

    public LiveData<Resource<User>> user = Transformations.switchMap(loginData, login -> {
        if (login == null) {
            return AbsentLiveData.Companion.create();
        }else{
            return mUserRepository.loadUser(login);
        }
    });

    public void setLogin(String login){
        if(!login.equals(loginData.getValue())){
            loginData.setValue(login);
        }
    }
}

File: app/src/main/java/com/jjf/template/ui/UserViewModel.kt
package com.jjf.template.ui

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.jjf.template.result.Resource
import com.jjf.template.result.User
import com.jjf.template.repository.UserRepository
import com.jjf.template.util.lifecycle.AbsentLiveData
import com.jjf.template.util.switchMap
import javax.inject.Inject

/**
 * @author xj
 * date: 18-11-8
 * description :
 */

class UserViewModel
@Inject constructor(private val mUserRepository: UserRepository) : ViewModel() {
    private val loginData = MutableLiveData<String>()
    val user:LiveData<Resource<User>> = loginData.switchMap { login->
        if(login == null){
            AbsentLiveData.create()
        }else{
            mUserRepository.loadUser(login)
        }
    }
    fun setLogin(login: String) {
        if (login != loginData.value) {
            loginData.value = login
        }
    }
}
